### DISCUSSION TITLE
Start recurring jobs multiple times?

### DISCUSSION THREAD

**Discussion by timmmerlan** [jobrunr/jobrunr#755] [General]
Hi there,
currently the `processRecurringJobs` function starts a recurring job multiple times if it was to be started more than once since its last execution. Up to now, I only had cases where it doesn't matter if the (same) recurring job is then started only once or multiple times in parallel. Is there a specific reason/use case where it makes sense to start multiple instances of the same job in parallel?
Thank you for your help!
Best regards,
Tim

**Comment by rdehuyss** [comment]
I'm not completely understanding your question to be honest. JobRunr will by default not allow multiple job instances from the same recurring job.

**Comment by selcukholep** [comment]
Hi,

I encountered the same error, but I realized that if the poll interval is greater than your recurring period, this situation can occur. Perhaps you are experiencing the same issue. 🤔

**Comment by pavloff-de** [comment]
That is exactly the problem that worries me. There are other reasons for multiple jobs as well. Method `processRecurringJobs()` catches them all and schedules for execution:
```
...
} else if(jobsToSchedule.size() > 1) {
    LOGGER.info("Recurring job '{}' resulted in {} scheduled jobs. This means a long GC happened and JobRunr is catching up.", recurringJob.getJobName(), jobsToSchedule.size());
    allJobsToSchedule.addAll(jobsToSchedule); // RUN ALL NOW ?
} else if(isAlreadyScheduledEnqueuedOrProcessing(recurringJob)) {
...
```
If long GC happened or Server hung for some reason or even if the wrong periods were selected ...

**Reply by rdehuyss** [reply]
Well, this is identical behaviour to other schedulers. See also #448.

**Comment by sajawaji** [comment]
I encountered the same error but it is not related to pollInterval. This is happening when the cron expression is `*/59 * * * * *`.  After digging deeper I found that the issue is the way JobRunr decides how many times a jobs needs to be run based on following method in `RecurringJob.toScheduledJobs` .

It is the way how nextRun value is calculated. 
For example :
if from time is at `50th` second of a minute and based on poll interval `upTo` value will be `5th` second of next minute. So we calculate `nexRun` value using `from` which is less than that of `upTo` value and one entry is added, again it calculates based on `nextRun` value that is calculated earlier. This time it calculated `nextRun`  value of `59th` second of minute which is still less than the `upTo` value. So it adds another entry and runs 2 instances of same job.

This is just one of the scenario I have observed. If necessary I can attach reproducible code.

**Reply by rdehuyss** [reply]
As a question, what do you think `*/59 * * * * *` means?

How many times per minute will it run do you think?

**Reply by sajawaji** [reply]
It should run every 59 seconds, that means it will run twice a minute. 
But the jobs run in parallel which is not expected according to JobRunr.
Also it some times schedules more than 2 instances.

**Reply by rdehuyss** [reply]
Guess again... it will run at:
00:00:59
00:01:00
00:01:59
00:02:00

and so on.

Please see [CronDrive](https://www.crondrive.com/test-cron-expression?utf8=%E2%9C%93&expression=*%2F59+*+*+*+*+*&time_zone=UTC&commit=Test) and Wikipedia:

> **Non Standard Character: Slash (/)**
> Note that frequencies in general cannot be expressed; only step values which evenly divide their range express accurate frequencies (for minutes and seconds, that's /2, /3, /4, /5, /6, /10, /12, /15, /20 and /30 because 60 is evenly divisible by those numbers; for hours, that's /2, /3, /4, /6, /8 and /12); all other possible "steps" and all other fields yield inconsistent "short" periods at the end of the time-unit before it "resets" to the next minute, second, or day; for example, entering */5 for the day field sometimes executes after 1, 2, or 3 days, depending on the month and leap year; this is because cron is stateless (it does not remember the time of the last execution nor count the difference between it and now, required for accurate frequency counting—instead, cron is a mere pattern-matcher).

**Reply by sajawaji** [reply]
Thanks.
